### PR TITLE
Pull and Push promises resolve to the registry digest of the image

### PR DESCRIPTION
**Pull request by afitzek**
Change-Type: major
Connects-To: #39
Signed-off-by: Andreas Fitzek <andreas@resin.io>
---- Autogenerated Waffleboard Connection: Connects to #39

### CODE REVIEW

File: index.coffee [hunk 1]
```diff
@@ -296,3 +305,16 @@ exports.DockerProgress = class DockerProgress
 	# "image" is a string, the name of the docker image
 	getImageLayerSizes: (image) ->
 		legacy.getImageLayerSizes(@docker, image)
+
+	# Extracts the digest value of an image from docker evnets
```
Comment (CameronDiver) [inline index.coffee]: `evnets` -> `events`

**Review by CameronDiver — APPROVED** [review]
There is also a typo in the commit, other than that, looks good!

**Review by dfunckt — COMMENTED** [review]
How is this going to be used? Can't we just resolve the promise with the digest?

**Review by CameronDiver — APPROVED** [review]

File: index.coffee [hunk 2]
```diff
@@ -12,6 +12,28 @@ exports.RegistryV2 = legacy.RegistryV2
 LEGACY_DOCKER_VERSION = '1.10.0'
 DEFAULT_PROGRESS_BAR_STEP_COUNT = 50
 
+tryExtractDigestHash = (evt) ->
+	return null if !evt?
```
Comment (dfunckt) [inline index.coffee]: This is redundant -- it's internal function and the way it's used guarantees it's won't ever receive null.

File: index.coffee [hunk 3]
```diff
@@ -12,6 +12,28 @@ exports.RegistryV2 = legacy.RegistryV2
...
+	if _.isString(evt.status)
+		matchPull = evt.status.match(/^Digest:\s([a-zA-Z0-9]+:[a-f0-9]+)$/)
+		if matchPull?
+			return matchPull[1]
+
+# Extracts the digest value of an image from docker events
+# for push and pull operations, if no digest value is found
+# null is returned
+extractDigestHash = (stream) ->
+	if _.isArray(stream)
```
Comment (dfunckt) [inline index.coffee]: This is redundant as well.

File: index.d.ts [hunk 4]
```diff
@@ -15,14 +15,14 @@ declare module 'docker-progress' {
 			onProgress: ProgressCallback,
 			options: any,
 			callback?: ((err: Error, data: any) => void),
-		): Promise<any>;
+		): Promise<string>;
```
Comment (dfunckt) [inline index.d.ts]: This is `string | null`

**Review by dfunckt — CHANGES_REQUESTED** [review]

**Review by dfunckt — APPROVED** [review]

**Pull request merged**